### PR TITLE
archive and delete the problem

### PR DESCRIPTION
**Pull request by Wally5077**
resolve : #109  
resolve : #150 

archive:
image

restore & delete:
image

### CODE REVIEW

File: src/services/ProblemService.js [hunk 1]
```diff
@@ -38,7 +39,14 @@ export class ProblemService extends AbstractService {
     }
 
     async getAllProblems() {
-        return this.axios.get(`/api/problems`).then(res => res.data);
+        return this.axios.get(`/api/problems`).then(res => res.data.map(obj => new Problem(obj)));
     }
```
Comment (nyngwang) [inline src/services/ProblemService.js:43]: NIT: how about move `.then(...)` on the next line?

File: src/app/problem/ProblemEditor.js [hunk 2]
```diff
@@ -25,12 +25,14 @@ const ProblemEditor = () => {
     const [editingState, setEditingState] = useState(false);
     const [problemDescription, setProblemDescription] = useState(undefined);
     const [lastProblemDescription, setLastProblemDescription] = useState(problemDescription);
+    const [problemArchived, setProblemArchived] = useState(false);
```
Comment (nyngwang) [inline src/app/problem/ProblemEditor.js:28]: I think `problemIsArchived`, `setProblemIsArchived` are more readable.

**Review by nyngwang — CHANGES_REQUESTED** [review]
Review Done.

File: src/app/problem/ProblemEditor.js (hunk 2 above)
Comment (Johnny850807) [reply src/app/problem/ProblemEditor.js:28]: Improvement: `isProblemArchived`, `setIsProblemArchived`

File: src/app/problem/ProblemEditor.js [hunk 3]
```diff
@@ -53,6 +55,41 @@ const ProblemEditor = () => {
...
+                          borderColor={"#A2A3B1"}
+                          buttonColor={"#FFFFFF"} fontColor={"#A2A3B1"}
+                          onClick={() => {
+                              problemService.archiveOrDeleteProblem(problemId)
+                                  .then(() => fetchProblem(problemId));
+                          }}/>
+        )
+    }
+
+    const RestoreAndDeleteProblemButtons = () => {
```
Comment (Johnny850807) [inline src/app/problem/ProblemEditor.js:71]: rename `RestoreOrDeleteProblemButtons`

File: src/models/Problem.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        this.languageEnvs = languageEnvs;
+        this.judgeMatchPolicyPluginTag = judgeMatchPolicyPluginTag;
+        this.judgeFilterPluginTags = judgeFilterPluginTags;
+        this.tags = tags;
+        this.testcases = testcases;
+        this.totalGrade = totalGrade;
+        this.visible = visible;
+        this.archived = archived;
+    }
+}
```
Comment (Johnny850807) [inline src/models/Problem.js:16]: Appreciated

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done!

File: src/app/problem/ProblemEditor.js (hunk 3 above)
Comment (Wally5077) [reply src/app/problem/ProblemEditor.js:71]: 這邊是同時有兩個按鈕的關係

image

Comment (Johnny850807) [reply src/app/problem/ProblemEditor.js:71]: Make sense.

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Review by nyngwang — APPROVED** [review]
LGTM.

**Pull request merged**